distributed/semaphore.py

```diff
@@ -0,0 +1,264 @@
+                raise ValueError(
+                    "Inconsistent max leases: %s, expected: %s"
+                    % (max_leases, self.max_leases[name])
+                )
+
+    @gen.coroutine
+    def _get_lease(self, client, name, identifier):
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired release immediately
+        while client not in self.scheduler.clients:
+            yield
+        with (yield self.locks[name].acquire()):
+            result = True
+            if len(self.leases[name]) < self.max_leases[name]:
+                self.leases[name].append(identifier)
+                self.leases_per_client[client][name].append(identifier)
+            else:
+                result = False
+            raise gen.Return(result)
+
+    @gen.coroutine
+    def acquire(
+        self, stream=None, name=None, client=None, timeout=None, identifier=None
+    ):
+        with log_errors():
+            if isinstance(name, list):
+                name = tuple(name)
+            w = _Watch(timeout)
+            w.start()
+
+            while True:
+                # Reset the event and try to get a release. The event will be set if the state
+                # is changed and helps to identify when it is worth to retry an acquire
+                self.events[name].clear()
+                future = self._get_lease(client, name, identifier)
+                if timeout is not None:
+                    future = gen.with_timeout(w.leftover(), future)
+                try:
+                    result = yield future
+                except gen.TimeoutError:
+                    result = False
+
+                # If acquiring fails, we wait for the event to be set, i.e. something has
+                # been released and we can try to acquire again (continue loop)
+                if not result:
+                    future = self.events[name].wait()
+                    if timeout is not None:
+                        future = gen.with_timeout(w.leftover(), future)
+                    try:
+                        yield future
+                        continue
+                    except gen.TimeoutError:
+                        result = False
+                raise gen.Return(result)
+
+    @gen.coroutine
+    def release(self, stream=None, name=None, client=None, identifier=None):
+        with log_errors():
+            if isinstance(name, list):
+                name = tuple(name)
+            if name in self.leases and identifier in self.leases[name]:
+                self.scheduler.loop.add_callback(
+                    self._release_value, name, client, identifier
+                )
+            else:
+                raise ValueError("Semaphore released too many times.")
+
+    @gen.coroutine
+    def _release_value(self, name, client, identifier):
+        with (yield self.locks[name].acquire()):
```

Comment:
Assuming that these locks aren't necessary (I don't think they're ever used around code with yields) then my guess is that a bit of this can be simplified.

Reply:
Yes, works without locks. Will remove them again.